Muhammad Ichsan Natawijaya

icanq opened the pull request:
# Tasks
- [x] Api-doc
- [x] Model User -> Admin | User -> username, role, email, password
- [x] Model Product -> Name, ImageUrl, Price, Stock
- [x] CRUD Product
> sesuai https://docs.google.com/document/d/1eqHnrt8CmNMVmcO9MxSY78iFtj89l_2-nYx5K7Z2FJI

Commit:
first init

Commit:
first initiate server

Commit:
sequelize init

Commit:
create User model & validation

Commit:
add Product model & validation

Commit:
create association between product & user

Commit:
generate seed for admin also use bcryptjs for handling password

Commit:
test success login

Commit:
Failed test cases for login
> wrong password
> email not registered
> wrong email/password

Commit:
fix beforeAll & afterAll user.test.js

dhimashary commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,42 @@
+
+//admin
+async function authorize(req, res, next) {
+  const UserId = +req.loggedin.id;
```
kenapa ga langsung aja if req.loggedin.role === "admin"

dhimashary commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,42 @@
+
+//customer
+async function authorizeCustomer(req, res, next) {
+  const UserId = +req.loggedin.id;
```
kenapa ga langsung aja if req.loggedin.role === "customer"

dhimashary commented on controllers/userController.js:
```diff
@@ -0,0 +1,62 @@
+  static login(req, res, next) {
+    User.findOne({
+      where: {
+        email: req.body.email,
```
tambahin di where, role: "admin", jadi nanti ada function yg berbeda untuk login customer dengan where nya role customer

dhimashary commented on __tests__/user.test.js:
```diff
@@ -0,0 +1,105 @@
+        done();
+      });
+  });
+
```
tambahin juga kasus email sama passwordnya customer tapi mencoba login sbg admin

Commit:
TDD Create

Commit:
TDD ERROR DONE
WRONG ACCESS TOKEN
DONT HAVE ACCESS_TOKEN
REQUIRED FIELD EMPTY
PRICE SET TO MINUS
STOCK SET TO MINUS
PRICE AND STOCK INPUT ARE STRING

Commit:
TDD GET

Commit:
TDD UPDATE

Commit:
TDD ERROR FOR UPDATE
WRONG ACCESS_TOKEN
STOCK SET TO MINUS
PRICE SET TO MINUS
PRICE & STOCK INPUTTED WITH STRING

Commit:
TDD DELETE
SUCCESS DELETE
NO ACCESS_TOKEN
NOT ADMIN

Commit:
remove unecessary lines

icanq commented:
```
SECRET_KEY=pergikebulan
PORT=4000
SALT=10
```

Commit:
added cors for client purpose

Commit:
fix admin password on seeding

Commit:
require dotenv

Commit:
fix server

Commit:
fix dotenv problem

dhimashary commented on controllers/userController.js:
```diff
@@ -0,0 +1,63 @@
+        } else if (compare(req.body.password, data.password)) {
+          const access_token = generateToken({
+            id: data.id,
+            email: data.email,
```
baiknya ketika login role nya juga di sign ke token ya

dhimashary commented on __tests__/product.test.js:
```diff
@@ -0,0 +1,439 @@
+      .post("/products")
+      .set({
+        access_token: access_token,
+        role: userRole,
```
baiknya ketika sign token rolenya dimasukin juga ke token yang dibuat, jadi ga di set header manual begini

dhimashary commented on __tests__/product.test.js:
```diff
@@ -0,0 +1,439 @@
+      .post("/products")
+      .set({
+        access_token: access_token,
+        role: "customer",
```
baiknya ketika sign token rolenya dimasukin juga ke token yang dibuat, jadi ga di set header manual begini

dhimashary commented on __tests__/product.test.js:
```diff
@@ -0,0 +1,439 @@
+      .post("/products")
+      .set({
+        access_token: access_token,
+        role: userRole,
```
baiknya ketika sign token rolenya dimasukin juga ke token yang dibuat, jadi ga di set header manual begini

dhimashary commented on __tests__/product.test.js:
```diff
@@ -0,0 +1,439 @@
+      .post("/products")
+      .set({
+        access_token: access_token,
+        role: userRole,
```
baiknya ketika sign token rolenya dimasukin juga ke token yang dibuat, jadi ga di set header manual begini

dhimashary commented on __tests__/product.test.js:
```diff
@@ -0,0 +1,439 @@
+      .post("/products")
+      .set({
+        access_token: access_token,
+        role: userRole,
```
baiknya ketika sign token rolenya dimasukin juga ke token yang dibuat, jadi ga di set header manual begini

dhimashary commented on __tests__/product.test.js:
```diff
@@ -0,0 +1,439 @@
+      .post("/products")
+      .set({
+        access_token: access_token,
+        role: userRole,
```
baiknya ketika sign token rolenya dimasukin juga ke token yang dibuat, jadi ga di set header manual begini

dhimashary commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,17 @@
+async function authorize(req, res, next) {
+  try {
+    const { role } = req.headers
```
role jangan dikirim lewat headers tapi dari req.loggedInUser.roles, kalo kaya gini gw customer bisa jadi admin dengan hanya kirim headers role

dhimashary commented:
api doc dimana ya can?

Commit:
initialize cart

Commit:
fix role auth
fixing by including role to the access_token
and check the role with extract it using jwt in the auth process

Commit:
crud & checkout
create cart
get  whats on user cart
update quanitity on the cart
delete cart
checkout cart

Commit:
fix fetch

Commit:
fix

Commit:
make it better

Commit:
fix server bagian auth

Commit:
fix association

Commit:
add apidoc
fix some trouble

Commit:
feat update pkg
